cmd/clusterctl/pkg/client/cluster/mover.go

```diff
@@ -119,32 +119,113 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
 	clusters := graph.getClusters()
 	o.log.Info("Clusters to move", "Count", len(clusters))
 
-	// Move Cluster by Cluster, ensuring all the dependent objects are moved as well and that the OwnerReferences relations are rebuilt in the target cluster.
-	// Nb. We are moving clusters one by one in sequential order for the sake of readability of the logs.
+	// Sets the pause field on the Cluster object in the source management cluster, so the controllers stop reconciling it.
 	for _, cluster := range clusters {
 		o.log.Info("Moving", "Cluster", cluster.identity.Name, "Namespace", cluster.identity.Namespace)
 
-		// Sets the pause field on the Cluster object in the source management cluster, so the controllers stop reconciling it.
 		if err := setClusterPause(o.fromProxy, cluster, o.log); err != nil {
 			return err
 		}
+	}
+
+	// Define the move sequence by processing the ownerReference chain, so we ensure that a node is moved only after its owners.
+	// Additionally nodes are grouped, so bulk of nodes can be moved in parallel. e.g.
+	// - All the Clusters should be moved first (group 1, processed in parallel)
+	// - All the MachineDeployments should be moved second (group 1, processed in parallel)
+	// - then all the MachineSets, then all the Machines, etc.
```

Comment:
If you're going to do it in this order, I think we may need to retain cluster.spec.paused=true while everything is moving. After the move is done, we can clear paused. Otherwise, if you have e.g. a Cluster, <Infra>Cluster, MachineDeployment, KubeadmConfigTemplate, <Infra>MachineTemplate, the MachineDeployment controller in the target cluster potentially will start creating new MachineSets, which will in turn create new Machines instead of reusing the ones that have yet to move. Is this concern valid?

Reply:
> The MachineDeployment controller in the target cluster potentially will start creating new MachineSets, which will, in turn, create new Machines instead of reusing the ones that have yet to move

@ncdc this was not the case in the former implementation of move/not observed in my simple tests.

However, if we prefer to retain cluster.spec.paused=true while everything is moving it should not be difficult to do and it puts us on the safe side.

The only downside I can think of is the fact that we will be forced to remove (or rethink) the "waitForMachine" check because it will never be satisfied before clearing paused, but if I remember well you already expressed some concerns on that...

Reply:
@detiber you have tons of pivot expertise - care to 👀?

Reply:
+1 to retaining `cluster.spec.paused=true` through the duration of the pivoting process, otherwise we are likely to end up with race conditions between the creation of resources for things like MachineDeployments/MachineSets.

Reply:
+1 from me as well, seems like the safest approach

Reply:
Ok, so I'm dropping MachineWait. Yay!